### PR TITLE
Pt 162537748 channels do not keep old txs

### PR DESCRIPTION
**Pull request by velzevur**
This PR closes channels' memory leaks. The code will be further improved:
* a proper GC is due
* the channels' GC shall not be executed on every off-chain tx

PT [SC: do not keep old transactions](https://www.pivotaltracker.com/story/show/162537748)

I will create separate PTs for the improvement bullet points above.

### CODE REVIEW

File: apps/aeutils/src/aeu_mtrees.erl [hunk 1]
```diff
@@ -102,6 +103,18 @@ new_with_backend(empty, DB) ->
 new_with_backend(<<_:256>> = Hash, DB) ->
     aeu_mp_trees:new(Hash, DB).
 
+-spec gc_old_nodes(mtree()) -> mtree().
+gc_old_nodes(Tree) ->
+    %% TODO: this is a workarond and not a proper GC
```
Comment (hanssv) [inline apps/aeutils/src/aeu_mtrees.erl]: This just re-builds the tree?! But that will skip all things no longer accessible I guess

File: apps/aechannel/src/aesc_offchain_state.erl [hunk 2]
```diff
@@ -132,14 +135,14 @@ check_reestablish_tx(SignedTx, State) ->
 check_update_tx(SignedTx, State, OnChainTrees, OnChainEnv, Opts) ->
     check_update_tx_(none, SignedTx, State, OnChainTrees, OnChainEnv, Opts).
 
-check_update_tx_(F, SignedTx, #state{signed_txs = Txs}=State, OnChainTrees,
-                 OnChainEnv, Opts) ->
+check_update_tx_(F, SignedTx, #state{signed_tx = OldSignedTx}=State, OnChainTrees,
```
Comment (hanssv) [inline apps/aechannel/src/aesc_offchain_state.erl]: No function clause that matches ?NO_TX?

File: apps/aeutils/src/aeu_mtrees.erl (hunk 1 above)
Comment (velzevur) [reply apps/aeutils/src/aeu_mtrees.erl]: Yes. A proper GC will be implemented.

File: apps/aechannel/src/aesc_offchain_state.erl (hunk 2 above)
Comment (velzevur) [reply apps/aechannel/src/aesc_offchain_state.erl]: Yes. All updates are based on preceding state. If there is no state - no update can be applied.

File: apps/aechannel/src/aesc_offchain_state.erl [hunk 3]
```diff
@@ -229,13 +233,13 @@ run_extra_checks(F, Mod, Tx) when is_function(F, 2) ->
 
 -spec add_signed_tx(aetx_sign:signed_tx(), state(), aec_trees:trees(),
                     aetx_env:env(), map()) -> state().
-add_signed_tx(SignedTx, #state{signed_txs=Txs0}=State, OnChainTrees,
+add_signed_tx(SignedTx, #state{}=State, OnChainTrees,
```
Comment (zp-sd) [inline apps/aechannel/src/aesc_offchain_state.erl]: It's not `add_signed_tx` anymore, but rather `set_signed_tx` now, but I guess it's easier for backwards compatibility 🙂

Comment (velzevur) [reply apps/aechannel/src/aesc_offchain_state.erl]: addressed

**Comment by velzevur** [comment]
Tracked PTs:
[SC: Optimize when the GC sweep](https://www.pivotaltracker.com/story/show/162611094)
[State trees GC](https://www.pivotaltracker.com/story/show/162611094)

**Comment by velzevur** [comment]
> test - Failed
but it failed to an error not seemengly relevant to the changes introduced by PR
```
%%% aecore_errorfree_SUITE ==> application_test: FAILED
%%% aecore_errorfree_SUITE ==> {{badmatch,[{dev2,{"epoch.log",
                   <<"2018-12-12 15:32:21.017 [error] <0.304.0> Supervisor exometer_report_logger_sup had child exometer_report_logger started with exometer_report_logger:start_link([{id,aec_metrics_logger},{input,[{mode,plugin},{module,aec_metrics},{state,[]}]},{output,[{mode,...},...]},...]) at <0.435.0> exit with reason {shutdown,{gen_server,call,[aec_metrics_rpt_dest,{check_destination,[ae,epoch,aecore,queues,sync_tasks,accepted,wait]}]}} in context child_terminated">>}}]},
 [{aecore_errorfree_SUITE,application_test,1,
                          [{file,"/home/builder/epoch/_build/test/lib/aecore/test/aecore_errorfree_SUITE.erl"},
                           {line,71}]},
  {test_server,ts_tc,3,[{file,"test_server.erl"},{line,1546}]},
  {test_server,run_test_case_eval1,6,[{file,"test_server.erl"},{line,1062}]},
  {test_server,run_test_case_eval,9,[{file,"test_server.erl"},{line,994}]}]}
%%% aecore_errorfree_SUITE ==> jobs_server_test: OK
```
restarted the job

**Pull request merged**